va/caaQuorum.go

```diff
@@ -0,0 +1,246 @@
+	apiResp, err := ctxhttp.Do(ctx, ic, req)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		_ = apiResp.Body.Close()
+	}()
+	body, err := ioutil.ReadAll(apiResp.Body)
+	if err != nil {
+		return nil, err
+	}
+	var respObj response
+	err = json.Unmarshal(body, &respObj)
+	if err != nil {
+		return nil, err
+	}
+	if respObj.Status != dns.RcodeSuccess {
+		if respObj.Comment != "" {
+			return nil, fmt.Errorf("Query failed with %s: %s", dns.RcodeToString[respObj.Status], respObj.Comment)
+		}
+		return nil, fmt.Errorf("Query failed wtih %s", dns.RcodeToString[respObj.Status])
+	}
+
+	return parseAnswer(respObj.Answer)
+}
+
+type queryResult struct {
+	records []*dns.CAA
+	err     error
+}
+
+type caaSet []*dns.CAA
+
+func (cs caaSet) Len() int           { return len(cs) }
+func (cs caaSet) Less(i, j int) bool { return cs[i].Value < cs[j].Value } // sort by value...?
+func (cs caaSet) Swap(i, j int)      { cs[i], cs[j] = cs[j], cs[i] }
+
+func hashCAASet(set []*dns.CAA) ([32]byte, error) {
+	var err error
+	offset, size := 0, 0
+	sortedSet := caaSet(set)
+	sort.Sort(sortedSet)
+	for _, rr := range sortedSet {
+		size += dns.Len(rr)
+	}
+	tbh := make([]byte, size)
+	for _, rr := range sortedSet {
+		ttl := rr.Hdr.Ttl
+		rr.Hdr.Ttl = 0 // only variable that should jitter
+		offset, err = dns.PackRR(rr, tbh, offset, nil, false)
+		if err != nil {
+			return [32]byte{}, err
+		}
+		rr.Hdr.Ttl = ttl
+	}
+	return sha256.Sum256(tbh), nil
+}
+
+// LookupCAA performs a multipath CAA DNS lookup using GPDNS
+func (cpr *CAAPublicResolver) LookupCAA(ctx context.Context, domain string) ([]*dns.CAA, error) {
+	query := make(url.Values)
+	query.Add("name", domain)
+	query.Add("type", "257") // CAA
+
+	results := make(chan queryResult, len(cpr.clients))
+	for addr, interfaceClient := range cpr.clients {
+		req, err := http.NewRequest("GET", apiURI, nil)
+		if err != nil {
+			return nil, err
+		}
+		req.URL.RawQuery = query.Encode()
+		go func(ic *http.Client, ia string) {
+			started := time.Now()
+			records, err := cpr.queryCAA(ctx, req, ic)
+			cpr.stats.TimingDuration(fmt.Sprintf("GPDNS.CAA.Latency.%s", ia), time.Since(started))
+			results <- queryResult{records, err}
+		}(interfaceClient, addr)
+	}
+	// collect everything
+	i := 0
+	failed := 0
+	var CAAs []*dns.CAA
+	var setHash [32]byte
+	var err error
+	for r := range results {
+		if r.err != nil {
+			failed++
+			if failed > cpr.maxFailures {
+				return nil, fmt.Errorf("%d out of %d CAA queries failed", len(cpr.clients), failed)
+			}
+		}
+		if CAAs == nil {
+			CAAs = r.records
+			setHash, err = hashCAASet(CAAs)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			hashedSet, err := hashCAASet(r.records)
+			if err != nil {
+				return nil, err
+			}
+			if len(r.records) != len(CAAs) || hashedSet != setHash {
+				return nil, errors.New("mismatching CAA record sets were returned")
+			}
+		}
+		i++
+		if i == len(cpr.clients) {
+			close(results) // break loop
```

Comment:
Do we get any special advantage by using `close` here rather than the slightly clearer `break`?